client/filejsonstore/filejsonstore.go

```diff
@@ -0,0 +1,153 @@
+package client
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/theupdateframework/go-tuf/client"
+	"github.com/theupdateframework/go-tuf/internal/fsutil"
+	"github.com/theupdateframework/go-tuf/util"
+)
+
+// ErrNotJSON is returned when a metadata operation is attempted
+// against a file that does not seem to be a JSON file
+// (e.g. does not end in .json, case sensitive).
+var ErrNotJSON = errors.New("file is not in JSON format")
+
+// ErrToPermissive is returned when the metadata directory, or a metadata
+// file has too permissive file mode bits set
+var ErrTooPermissive = errors.New("permissions are too permissive")
+
+const (
+	// user:  rwx
+	// group: r-x
+	// other: ---
+	dirCreateMode = os.FileMode(0750)
+	// user:  rw-
+	// group: r--
+	// other: ---
+	fileCreateMode = os.FileMode(0640)
+)
+
+// FileJSONStore represents a local metadata cache relying on raw JSON files
+// as retrieved from the remote repository.
+type FileJSONStore struct {
+	mtx     sync.RWMutex
+	baseDir string
+}
+
+var _ client.LocalStore = (*FileJSONStore)(nil)
+
+// NewFileJSONStore returns a new metadata cache, implemented using raw JSON
+// files, stored in a directory provided by the client.
+// If the provided directory does not exist on disk, it will be created.
+// The provided metadata cache is safe for concurrent access.
+func NewFileJSONStore(baseDir string) (*FileJSONStore, error) {
+	var f = FileJSONStore{
+		baseDir: baseDir,
+	}
+	var err error
+
+	// Does the directory exist?
+	fi, err := os.Stat(baseDir)
+	if err != nil {
+		pe, ok := err.(*os.PathError)
+		if ok && errors.Is(pe.Err, os.ErrNotExist) {
+			// Create the directory
+			if err = os.MkdirAll(baseDir, dirCreateMode); err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	} else {
+		// Verify that it is a directory
+		if !fi.IsDir() {
+			return nil, fmt.Errorf("can not open %s, not a directory",
+				baseDir)
+		}
+		// Verify file mode is not too permissive.
+		if err = fsutil.EnsurePermission(fi, dirCreateMode); err != nil {
+			return nil, ErrTooPermissive
+		}
+	}
+
+	return &f, nil
+}
+
+// GetMeta returns the currently cached set of metadata files.
+func (f *FileJSONStore) GetMeta() (map[string]json.RawMessage, error) {
+	f.mtx.RLock()
+	defer f.mtx.RUnlock()
+
+	names, err := os.ReadDir(f.baseDir)
+	if err != nil {
+		return nil, err
+	}
+
+	meta := map[string]json.RawMessage{}
+	for _, name := range names {
+		p := filepath.FromSlash(filepath.Join(f.baseDir, name.Name()))
```

Comment:
nit: can we put `p` closer to where it's used?

Also, is `filepath.FromSlash` necessary, given that `filepath.Join` already uses an OS-specific filepath separator?

Reply:
The issue is that we don't know what's inside `f.baseDir`, but that could be fixed in `NewFileJSONStore`. But as I read through the code of `Join` I found that prior to returning the value, `Clean` is called, and `Clean` calls `FromSlash`: https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/path/filepath/path.go;l=89

So yes, this `FromSlash` is redundant.